### ISSUE TITLE
"Uncaught RangeError: Maximum call stack size exceeded": deepEqual chokes on props.children

### ISSUE THREAD

**Issue by gabrielecirulli** [erikras/redux-form#919]
In `6.0.0-alpha.7`,  when using the new `<Field>` element, I've been encountering a strange error when trying to compose elements that accept `children` props and passing them their children via the `<Field>` component.

Here is the most reduced working example I could get:

```jsx
import React, { Component } from "react";
import { reduxForm, Field } from "redux-form/immutable";

export default class InnerComponent extends Component {
  render() {
    return (
      <div>
        Something
      </div>
    );
  }
}

export default class CustomField extends Component {
  render() {
    return (
      <div onClick={() => this.props.onChange(true)}>
        {this.props.children}
      </div>
    );
  }
}

@reduxForm({
  form: "event",
})
export default class EventForm extends Component {
  render() {
    return (
      <Field
        name="customField"
        component={CustomField}>
          <InnerComponent/>
      </Field>
    );
  }
}
```

In a nutshell, `CustomField` will render anything that's passed as children to it. It has a simple `onClick` handler to trigger the `onChange` callback passed by `redux-form`, which sets the whole issue in motion.

`InnerComponent` is passed as a child. In my experiments, this issue does not happen when you pass just text inside `<Field>`.

From what I was able to figure out, the following file is at the root of the stack overflow: https://github.com/erikras/redux-form/blob/v6/src/ConnectedField.js#L19

What it does is it performs a deep comparison between the previous props and the next props of `<ConnectedField/>`, which is in-between `<Field>`, `<CustomField>` and `<InnerComponent>`.

![set up a new event impraise 2016-05-06 6 pm-48-10](https://cloud.githubusercontent.com/assets/886011/15079840/2500f18c-13bb-11e6-97b5-7faeaafce945.png)

`<ConnectedField/>`'s `props.children` is an `InnerComponent` element. The deep comparison moves on to `InnerComponent`'s `_owner` (private React property), which in this case is the top-level `EventForm` component. 

From that point, it's extremely easy to identify a circular reference, as you would only have to follow the `props.children` chain back down to find `InnerComponent` in order to close the loop.

When this happens, the component fails to update.

Potential solutions that I can think of:

 - Explicitly skip comparing the `children` property in the props object. Potential drawback: the component won't be able to tell when props changed. Potential workaround: only perform a shallow comparison of `children` objects.
 - Do not consider underscore-prepended properties in the deep comparison. Potential drawback: very opinionated approach and may interfere with things that are not related to React.
 - Impose a hard limit on the depth of the comparison that won't be exceeded. If everything matches up to that point, the objects are considered equal. Potential drawback: the circular reference may start at any point in the chain, even in shallow areas, which would moot this check.

--------------------------------

I also attempted to pass the whole structure directly through the `component` prop:

```jsx
import React, { Component } from "react";
import { reduxForm, Field } from "redux-form/immutable";

export default class InnerComponent extends Component {
  render() {
    return (
      <div>
        Something
      </div>
    );
  }
}

export default class CustomField extends Component {
  render() {
    return (
      <div onClick={() => this.props.onChange(true)}>
        {this.props.children}
      </div>
    );
  }
}

@reduxForm({
  form: "event",
})
export default class EventForm extends Component {
  render() {
    return (
      <Field
        name="customField"
        component={props => <CustomField><InnerComponent/></CustomField>}>
          
      </Field>
    );
  }
}
```

In this case, I get the following error:

`ReactCompositeComponent.js:154 Uncaught TypeError: Can't add property context, object is not extensible`.

I did not investigate down this path further, as it seems like my initial attempt is more in line with the intended use of this API.

Secondary details:
 - I'm using React 0.13 and Babel 5

**Comment by erikras** [comment]
`deepEqual` needs to be rewritten. Thats on the soon-to-do list.

You're not using `redux-form/immutable`, right?

**Comment by gabrielecirulli** [comment]
I'm using the immutable flavour, yes. Also sorry, I just realised by seeing your reply that I accidentally submitted this before finishing it. I'm preparing a much more detailed write up of my findings. Please check back in a bit.

**Comment by gabrielecirulli** [comment]
I just updated this issue with a much more detailed write-up. Please let me know if this is any more helpful.

As this is a really stringent issue for me, I'd love to contribute a solution, but I'm not sure where to start. If you have an idea on how you would rewrite `deepEqual` to prevent this from happening, I would love to implement it if you share your thoughts with me.

Also, do you have more information about in which way the `immutable` version of `deepEqual` fails and what the difference would be between it and the vanilla JS version?

**Comment by gabrielecirulli** [comment]
I removed the dependency from `CustomField` in my first example by reducing it to simply use a `<button>`. This should bring even more clarity to help solve the issue:


```jsx
import React, { Component } from "react";
import { reduxForm, Field } from "redux-form/immutable";

export default class InnerComponent extends Component {
  render() {
    return (
      <div>
        Something
      </div>
    );
  }
}

@reduxForm({
  form: "event",
})
export default class EventForm extends Component {
  render() {
    return (
      <Field
        name="something"
        component="button">
          <InnerComponent/>
      </Field>
    );
  }
}
```

**Comment by erikras** [comment]
☝️ The code in your previous comment is all that it takes to cause the stack overflow? Just clarifying.

**Comment by gabrielecirulli** [comment]
Indeed, that seems to be the case on my end. Please give it a spin on your end as well if you get a chance.

**Comment by erikras** [comment]
I meant the "plain" `deepEqual` needs a rewrite. I'm not going to have time to look into this today.

If you want to get started fixing it, should see if you can recreate it in [`deepEqual.spec.js`](https://github.com/erikras/redux-form/blob/v6/src/structure/immutable/__tests__/deepEqual.spec.js), and then repair [`deepEqual.js`](https://github.com/erikras/redux-form/blob/v6/src/structure/immutable/deepEqual.js) to avoid the overflow.

**Comment by gabrielecirulli** [comment]
No worries, didn't mean to rush you. I'll look into how other libraries deal with cyclical objects. In case you have any idea for a strategy to prevent the overflow, I'd love to hear it.

**Comment by gabrielecirulli** [comment]
@erikras what are your considerations on adding a dependency on `lodash`? As I understand `redux-form` does not seem to depend on it, maybe for leanness reasons, however it seems like it might be a favourable choice. `redux` already depends on it, so any user of `redux-form` will be pulling in `lodash` anyway.

I believe `lodash` has a much more tried-and-true (and efficient) approach to object comparison, so it might be a great choice. For time-sensitivity reasons I am forking `redux-form` and replacing `deepEqual` with `lodash.isEqual`, until there's an official solution. 

If you're interested in pushing this through, I can open a PR with this change.

**Comment by erikras** [comment]
I would like to avoid lodash if possible, for the reasons discussed in #859.

**Comment by gabrielecirulli** [comment]
I see. [Here is my attempt at implementing the fix with lodash.](https://github.com/impraise/redux-form/compare/v6...fix-deep-equal-cyclical?diff=unified&name=fix-deep-equal-cyclical) The tests are green. It's fine if you don't consider it mergeable though. For now I just need something that's stable. If there's more time I will go back and have a stab at removing lodash, but otherwise I'll just depend on this until there's an official solution.

Thanks for your help so far!

**Comment by erikras** [comment]
Doh! 🙈 Can you PR this, @gabrielecirulli?

**Comment by gabrielecirulli** [comment]
Here you go: #929

**Comment by erikras** [comment]
Published as [`v6.0.0-alpha.8`](https://github.com/erikras/redux-form/releases/tag/v6.0.0-alpha.8).
